### PR TITLE
MAINT: adapt to array-api-strict 2.0

### LINKED ISSUE #21073: MAINT: `cluster`/`stats`: array API test failures in main
### Describe your issue.

The array API CI job is failing (e.g. [here](https://github.com/scipy/scipy/actions/runs/9712970168/job/26808785497)) with these errors, probably due to changes in array-api-strict.

### Reproducing Code Example

```python
NA
```

### Error message

```shell
FAILED scipy/cluster/tests/test_hierarchy.py::TestDendrogram::test_dendrogram_single_linkage_tdist[array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/cluster/tests/test_hierarchy.py::TestDendrogram::test_labels_as_array_or_list[array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/cluster/tests/test_hierarchy.py::TestDendrogram::test_dendrogram_leaf_colors_zero_dist[array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/cluster/tests/test_hierarchy.py::TestDendrogram::test_dendrogram_leaf_colors[array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circfuncs[circvar-0.006455174000787767-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circfuncs[circstd-6.520702116-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circfuncs_small[array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circfuncs_close[circvar-var-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circfuncs_close[circstd-std-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circmean_axis[circvar-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circmean_axis[circstd-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circfuncs_array_like[circvar-0.006455174270186603-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circfuncs_array_like[circstd-6.520702116-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_empty[circvar-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_empty[circstd-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_nan_propagate[circvar-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_nan_propagate[circstd-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_nan_propagate_array[circvar-expected1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_nan_propagate_array[circstd-expected2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circfuncs_uint8[array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_morestats.py::TestCircFuncs::test_circstd_zero[array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_input_validation[array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_axis_dtype[-3-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_axis_dtype[-2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_axis_dtype[-1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_axis_dtype[0-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_axis_dtype[1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_axis_dtype[2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_alternative[two-sided-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_alternative[less-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_resampling.py::TestMonteCarloHypothesisTest::test_alternative[greater-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_r_almost_exactly_pos1[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_r_almost_exactly_neg1[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_basic[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_constant_input[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_near_constant_input[float32-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_near_constant_input[float64-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_very_small_input_values[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_very_large_input_values[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_extremely_large_input_values[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_length_two_pos1[array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_length_two_neg1[array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_negative_correlation_pvalue_gh17795[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_length3_r_exactly_negative_one[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_nd_special_cases[array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_array_api[less-0-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_array_api[less-1-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_array_api[less-None-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_array_api[greater-0-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_array_api[greater-1-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_array_api[greater-None-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_array_api[two-sided-0-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_array_api[two-sided-1-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPearsonr::test_array_api[two-sided-None-array_api_strict] - RuntimeError: The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestMoments::test_moment_center_array_moment[None-10-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestMoments::test_moment_center_array_moment[None-size1-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestMoments::test_moment_center_array_moment[0-10-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestMoments::test_moment_center_array_moment[0-size1-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestMoments::test_moment_center_array_moment[1-10-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestMoments::test_moment_center_array_moment[1-size1-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestMoments::test_moment[array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat0-case1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat0-case2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat0-case3-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat1-case1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat1-case2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat1-case3-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat2-case1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat2-case2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat2-case3-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat3-case1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat3-case2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat3-case3-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat4-case1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat4-case2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat4-case3-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat5-case1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat5-case2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat5-case3-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat6-case1-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat6-case2-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_basic[lambda_stat6-case3-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_axis[array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_ddof_broadcasting[array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_power_divergence_gh_12282[array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestPowerDivergence::test_power_divergence_against_cressie_read_data[array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestChisquare::test_gh_chisquare_12282[array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestChisquare::test_chiquare_data_types_attributes[200-uint8-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestChisquare::test_chiquare_data_types_attributes[1000000-int32-array_api_strict] - RuntimeError: The function minimum requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12
FAILED scipy/stats/tests/test_stats.py::TestSkewTest::test_against_R[two-sided-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestSkewTest::test_against_R[less-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestSkewTest::test_against_R[greater-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestKurtosisTest::test_against_R[two-sided-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestKurtosisTest::test_against_R[less-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestKurtosisTest::test_against_R[greater-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestNormalTest::test_against_R[two-sided-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestNormalTest::test_against_R[less-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
FAILED scipy/stats/tests/test_stats.py::TestNormalTest::test_against_R[greater-array_api_strict] - TypeError: array iteration is not allowed in array-api-strict
```

### SciPy/NumPy/Python version and system information

```shell
NA
```

### PR DESCRIPTION
**Pull request by lucascolley**
Closes gh-21073.

Draft for now as some of the errors remain.

### CODE REVIEW

**Comment by lucascolley** [comment]
@asmeurer please could you explain how we can work around the `array iteration is not allowed in array-api-strict` failures in CI here?

**Comment by mdhaber** [comment]
I think typically the solution is to use integer indices rather than iterating directly over zeroth axis slices or unpacking the array as if it were a tuple.

**Comment by j-bowhay** [comment]
> I'm surprised these things weren't caught before, though. I thought array-api-strict arrays never supported unpacking, for instance.

I think this was because of the bug I mentioned at the summit which has now been fixed in https://github.com/data-apis/array-api-strict/pull/42

**Comment by mdhaber** [comment]
@j-bowhay Wow I didn't realize it was so widespread. 

@lucascolley were you going to include `array iteration is not allowed in array-api-strict` fixes here? I'd be happy to review.

**Comment by lucascolley** [comment]
@mdhaber yep, currently busy moving away from uni, will be done in the coming days

**Comment by asmeurer** [comment]
You can also use the new `unstack` function.

**Comment by asmeurer** [comment]
I'm actually considering whether this change in array-api-strict was too far. From a technical standpoint, the array API could allow `__iter__` on 1-D arrays, because it conforms to the Python default `object.__iter__` implementation. At the same time, there were explicit reasons for not including `__iter__` in the array API (see https://github.com/data-apis/array-api/issues/188), so morally array-api-strict is following the consortium by disallowing it.

File: scipy/cluster/tests/test_hierarchy.py [hunk 1]
```diff
@@ -948,7 +948,11 @@ def test_labels_as_array_or_list(self, xp):
         Z = linkage(xp.asarray(hierarchy_test_data.ytdist), 'single')
         labels = xp.asarray([1, 3, 2, 6, 4, 5])
         result1 = dendrogram(Z, labels=labels, no_plot=True)
-        result2 = dendrogram(Z, labels=list(labels), no_plot=True)
+        # imitating `tolist` in a standard-compliant way
+        labels_list = []
+        for i in range(labels.shape[0]):
+            labels_list += [int(labels[i])]
+        result2 = dendrogram(Z, labels=labels_list, no_plot=True)
```
Comment (mdhaber) [inline scipy/cluster/tests/test_hierarchy.py:955]: @asmeurer regarding your [comment](https://github.com/scipy/scipy/pull/21074#issuecomment-2197772042): yes, this sort of change is unfortunate.

File: scipy/cluster/tests/test_hierarchy.py [hunk 2]
```diff
@@ -948,7 +948,11 @@ def test_labels_as_array_or_list(self, xp):
         Z = linkage(xp.asarray(hierarchy_test_data.ytdist), 'single')
         labels = xp.asarray([1, 3, 2, 6, 4, 5])
```
Comment (mdhaber) [inline scipy/cluster/tests/test_hierarchy.py:949]:
```suggestion
        labels = [1, 3, 2, 6, 4, 5]
```

File: scipy/cluster/tests/test_hierarchy.py [hunk 3]
```diff
@@ -948,7 +948,11 @@ def test_labels_as_array_or_list(self, xp):
...
         result1 = dendrogram(Z, labels=labels, no_plot=True)
```
Comment (mdhaber) [inline scipy/cluster/tests/test_hierarchy.py:950]:
```suggestion
        result1 = dendrogram(Z, labels=xp.asarray(labels), no_plot=True)
```

File: scipy/cluster/tests/test_hierarchy.py (hunk 1 above)
Comment (mdhaber) [inline scipy/cluster/tests/test_hierarchy.py:955]: How about this @lucascolley?
```suggestion
        result2 = dendrogram(Z, labels=labels, no_plot=True)
```

Comment (lucascolley) [reply scipy/cluster/tests/test_hierarchy.py:955]: yep, much nicer

File: scipy/conftest.py [hunk 4]
```diff
@@ -117,6 +117,9 @@ def check_fpu_mode(request):
     try:
         import array_api_strict
         xp_available_backends.update({'array_api_strict': array_api_strict})
+        array_api_strict.set_array_api_strict_flags(
+            api_version='2023.12'
+        )
```
Comment (mdhaber) [inline scipy/conftest.py:122]: Locally this gave me an error before I upgraded to 2.0. Might want to let everyone know to upgrade.

File: scipy/stats/tests/test_stats.py [hunk 5]
```diff
@@ -6397,7 +6397,8 @@ def test_against_R(self, alternative, xp):
         # test_result@test$p.value
         test_name = self.test_name
         test_fun = getattr(stats, test_name)
-        ref_statistic, ref_pvalue = xp.asarray(self.case_ref)
+        xp_case_ref = xp.asarray(self.case_ref)
+        ref_statistic, ref_pvalue = xp_case_ref[0], xp_case_ref[1]
```
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:6401]: Just for consistency with what was done above. Could go either way... this way is probably less efficient, but doesn't need an intermediate variable. 
```suggestion
        ref_statistic= xp.asarray(self.case_ref[0])
        ref_pvalue = xp.asarray(self.case_ref[1])
```
I suppose this create NumPy 0d arrays whereas the one above is NumPy scalars, but doesn't seem to matter.

File: scipy/stats/_stats_py.py [hunk 6]
```diff
@@ -1079,11 +1079,12 @@ def moment(a, order=1, axis=0, nan_policy='propagate', *, center=None):
         calculate_mean = center is None and xp.any(order > 1)
         mean = xp.mean(a, axis=axis, keepdims=True) if calculate_mean else None
         mmnt = []
-        for i in order:
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:1082]: I imagine you prefer the more explicit variable name, so I'll leave it, but to reduce the diff:
```python3
        for i in (order[j] for j in range(order.shape[0])):
```

**Review by mdhaber — APPROVED** [review]

File: scipy/conftest.py (hunk 4 above)
Comment (lucascolley) [reply scipy/conftest.py:122]: we could add a version check and a more helpful error message here, and add a lower bound to `pyproject.toml`?

Comment (mdhaber) [reply scipy/conftest.py:122]: Ah, sounds good.

Comment (lucascolley) [reply scipy/conftest.py:125]: Okay, I tried to do this, please test

**Comment by fancidev** [comment]
> I'm actually considering whether this change in array-api-strict was too far. From a technical standpoint, the array API could allow `__iter__` on 1-D arrays, because it conforms to the Python default `object.__iter__` implementation. At the same time, there were explicit reasons for not including `__iter__` in the array API (see [data-apis/array-api#188](https://github.com/data-apis/array-api/issues/188)), so morally array-api-strict is following the consortium by disallowing it.

I think there is some merit in disallowing iteration over an array. Such iteration fits the conceptual model that an array is a sequence of rows (of collection of rows (of ….)). This may not work very nicely with column-major arrays. There are even some cases (e.g. pandas DataFrame) where the default iteration is over columns.

So I think the [unstack](https://data-apis.org/array-api/latest/API_specification/generated/array_api.unstack.html) function that @asmeurer mentions is a viable alternative that balances readability and explicitness.

**Comment by asmeurer** [comment]
Ok I might reenable iteration for 1-D arrays. You might want to hold off on merging this until then.

**Comment by lucascolley** [comment]
Given this is making CI red everywhere, I propose we merge and we can always circle back to usage of `unstack` later.

**Comment by asmeurer** [comment]
https://github.com/data-apis/array-api-strict/pull/48

**Comment by asmeurer** [comment]
I've just released array-api-strict 2.0.1, which re-allows iteration on 1-D arrays. Many of the changes here can likely be reverted.

**Pull request merged**